### PR TITLE
Use pluck

### PR DESCRIPTION
**Pull request by andycroll**
Bit late! Sorry!

### CODE REVIEW

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+layout: article
+category: ruby
+image:
+  base: '2018/using-pluck-can-save-a-bunch-of-memory'
+  alt: 'Guitarist'
+  source: 'https://unsplash.com/photos/DUrdHDn81mk'
+  credit: 'Jacek Dylag'
+---
+
+Active Record models are incredible flexible and provide a very large amount of flexibility. One of the consequences of providing this large API of methods is that each individual Active Record object takes up a lot of space when loaded into memory.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: Choose one of the statements re flexibility. 😁

Comment (nodunayo) [reply _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: *One of the consequence of having a large API of methods is that....

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+
+If you only use a single field, or just need the `id`s, Rails has your back.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: Why not? "If you only use a single field, for example, the `id`s, then Rails has your back."?

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+Book.paperbacks.pluck(:title)
+#=> ["Eloquent Ruby", "Sapiens", "Agile Web Development With Rails"]
+```
+
+
+## But why?
+
+This is about speed and efficient memory usage.
+
+Active Record models can take up a lot of memory, particularly true if you're operating over large collections. Each row from the database is transformed into an object and then in the first example we're just picking a single field, you can see that all the extra functionality that a model has (and Rails prepares for you in memory) is wasted.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: *Active Record models can take up a lot of memory and this becomes particularly painful/onerous, etc..if you're operating over large collections.

Comment (nodunayo) [reply _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: *In the first example, each row from the database is transformed into an object, and then we're only picking a single field. All the extra functionality that a model has — and that Rails prepares for you in memory — is wasted.

Comment (nodunayo) [reply _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: Maybe we should change 'wasted' to something like 'not needed' or 'superfluous' or 'unnecessary'. Because, 'wasted' implies you could and should use it but you just don't — whereas in the example, they don't need the extra stuff at all for their purposes.

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+
+Just picking the fields you need means your application will run faster and use less memory.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: *Only loading the fields that you need means your application...

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+
+
+### Why not?
+
+The `#pluck` method returns only the values you request in a plain Ruby array, no methods are loaded from your model.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: Semicolon or em-dash instead of comma after "plain Ruby array".

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 6]
```diff
@@ -0,0 +1,59 @@
...
+
+If you actually need the actual Active Record model or need to update at some point after using `#pluck` you’re out of luck.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: *If you do need the actual Active Record model or need to update it at some point after....`#pluck`, (comma) you're out of luck.

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 7]
```diff
@@ -0,0 +1,59 @@
...
+
+You might also consider using a `select` clause in your scope, although this is still not as memory efficient as using `#pluck`.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: Why would I choose this over `#pluck`?

**Review by nodunayo — DISMISSED** [review]
Not bad. Mainly some copy changes. Structurally it's there, I reckon.

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 8]
```diff
@@ -0,0 +1,59 @@
...
+layout: article
+category: ruby
+image:
+  base: '2018/using-pluck-can-save-a-bunch-of-memory'
+  alt: 'Guitarist'
+  source: 'https://unsplash.com/photos/DUrdHDn81mk'
+  credit: 'Jacek Dylag'
+---
+
+Active Record models are incredible flexible and provide a very large amount of functionality. One consequence of providing this large API of methods is that each individual Active Record object takes up a lot of space when loaded into memory.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: *incredibly

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+If you only use a single field, for example, the `id`s, then Rails has your back.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: *If you only need to use a single field on an Active Record model...

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+Book.paperbacks.map { |book| book.title }
+#=> ["Eloquent Ruby", "Sapiens", "Agile Web Development With Rails"]
+Book.paperbacks.map(&:title)
+#=> ["Eloquent Ruby", "Sapiens", "Agile Web Development With Rails"]
+```
+
+
+## Use…
+
+...the `ActiveRelation#pluck` method to pull the fields directly from the database.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: *pull the _required_ fields

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+Book.paperbacks.pluck(:title)
+#=> ["Eloquent Ruby", "Sapiens", "Agile Web Development With Rails"]
+```
+
+
+## But why?
+
+This is about speed and efficient memory usage.
+
+Active Record models can take up a lot of memory and this becomes particularly painful if you're operating over large collections. In the first example, each row from the database is transformed into an object and we're only picking a single field. All the extra functionality that a model has—and Rails prepares for you in memory—is unneeded.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: You've already said they can take up a lot of memory in the beginning. Therefore, how about rewording this to something like:

"The problem of Active Record models taking up a lot of memory becomes particularly painful if..." or "The fact that Active Record models can take up a lot of memory becomes particularly painful if...".

Comment (nodunayo) [reply _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: *In the first example, each row from the database is transformed into an object and _yet_ we're only picking _out_/_using_ a single field.

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 12]
```diff
@@ -0,0 +1,59 @@
...
+
+Only loading the fields that you need means your application will run faster and use less memory.
+
+
+### Why not?
+
+The `#pluck` method returns only the values you request in a plain Ruby array—no methods are loaded from your model.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md:50]: Add spacing around the hyphen. Also, technically, it should be an em-dash: —. :P

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md (hunk 11 above)
Comment (nodunayo) [reply _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: Add spaces around the hyphens...

Convert them to em-dashes if you care about being technically correct. :P

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 13]
```diff
@@ -0,0 +1,59 @@
...
+
+If you do need the actual Active Record model or need to update at some point after using `#pluck`, you’re out of luck.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: Do you mean update the model? If so, make that clearer here.

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md [hunk 14]
```diff
@@ -0,0 +1,59 @@
...
+
+You might also consider using a `select` clause in your scope, which would provide you Active Record functionality on the fields you request. This is still not as memory efficient as using `#pluck`.
```
Comment (nodunayo) [inline _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: Yess — way better.

*which would provide you with _the_ Active Record functionality but only on the fields you request. 

(Though, thinking about it, I'm not sure what functionality that comes with. Do you mean the return result is not a plain ol' array, with an AR object?)

**Review by nodunayo — DISMISSED** [review]
Almost there!

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md (hunk 11 above)
Comment (andycroll) [reply _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: They are em-dashes. Unicode. 😸

Comment (nodunayo) [reply _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md]: Alrighty! 😁

File: _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md (hunk 12 above)
Comment (andycroll) [reply _posts/2018/2018-09-16-pluck-can-save-a-bunch-of-memory.md:50]: If we're gonna out-typography nerd each other... this will become an epic struggle 😄 `—` == —
